File: packages/cli/tests/run.test.ts
import { describe, it, expect } from "vitest";
import { execaNode } from "../src/test-helpers.js";

describe("run command", () => {
  it("run appears in default help", async () => {
    const { stdout, exitCode } = await execaNode(["--help"]);
    expect(exitCode).toBe(0);
    expect(stdout).toContain("run");
  });

  it("run appears in --help-maturity", async () => {
    const { stdout, exitCode } = await execaNode(["--help-maturity"]);
    expect(exitCode).toBe(0);
    expect(stdout).toContain("run");
    expect(stdout).toContain("beta:");
  });

  it("run --list outputs builtin:ci", async () => {
    const { stdout, exitCode } = await execaNode(["run", "--list"]);
    expect(exitCode).toBe(0);
    expect(stdout).toContain("builtin:ci");
  });

  it("run --list --json outputs parseable recipes", async () => {
    const { stdout, exitCode } = await execaNode(["run", "--list", "--json"]);
    expect(exitCode).toBe(0);
    const result = JSON.parse(stdout);
    expect(result.recipes).toBeInstanceOf(Array);
    expect(result.recipes).toContain("builtin:ci");
  });

  it("run builtin:ci --dry-run outputs planned steps", async () => {
    const { stdout, exitCode } = await execaNode([
      "run",
      "builtin:ci",
      "--dry-run",
    ]);
    expect(exitCode).toBe(0);
    expect(stdout).toContain("doctor");
    expect(stdout).toContain("examples_verify");
    expect(stdout).toContain("verify_ci_standard");
  });

  it("run builtin:ci --dry-run --json outputs parseable result", async () => {
    const { stdout, exitCode } = await execaNode([
      "run",
      "builtin:ci",
      "--dry-run",
      "--json",
    ]);
    expect(exitCode).toBe(0);
    const result = JSON.parse(stdout);
    expect(result.recipe).toBe("builtin:ci");
    expect(result.ok).toBe(true);
    expect(result.steps).toBeInstanceOf(Array);
    expect(result.steps.length).toBeGreaterThan(0);
  });

  it("run unknown recipe exits with error", async () => {
    const { stderr, exitCode } = await execaNode(["run", "unknown"]);
    expect(exitCode).toBe(2);
    expect(stderr).toContain("unknown recipe");
  });

  it("run --help shows usage", async () => {
    const { stdout, exitCode } = await execaNode(["run", "--help"]);
    expect(exitCode).toBe(0);
    expect(stdout).toContain("run");
  });
});

File: internal/cli/run_test.go
package cli

import (
	"bytes"
	"encoding/json"
	"strings"
	"testing"
)

func TestRunList(t *testing.T) {
	var stdout bytes.Buffer
	var stderr bytes.Buffer
	code := Run([]string{"run", "--list"}, &stdout, &stderr)
	if code != ExitOK {
		t.Fatalf("expected exit code %d, got %d. stderr: %s", ExitOK, code, stderr.String())
	}
	out := stdout.String()
	if !strings.Contains(out, "builtin:ci") {
		t.Fatalf("expected builtin:ci in list, got: %s", out)
	}
}

func TestRunListJSON(t *testing.T) {
	var stdout bytes.Buffer
	var stderr bytes.Buffer
	code := Run([]string{"run", "--list", "--json"}, &stdout, &stderr)
	if code != ExitOK {
		t.Fatalf("expected exit code %d, got %d. stderr: %s", ExitOK, code, stderr.String())
	}
	var result map[string]any
	if err := json.Unmarshal(stdout.Bytes(), &result); err != nil {
		t.Fatalf("expected valid JSON: %v\noutput: %s", err, stdout.String())
	}
	recipes, ok := result["recipes"].([]any)
	if !ok || len(recipes) == 0 {
		t.Fatalf("expected recipes array, got: %+v", result)
	}
	found := false
	for _, r := range recipes {
		if r == "builtin:ci" {
			found = true
			break
		}
	}
	if !found {
		t.Fatalf("expected builtin:ci in recipes, got: %+v", result)
	}
}

func TestRunDryRun(t *testing.T) {
	var stdout bytes.Buffer
	var stderr bytes.Buffer
	code := Run([]string{"run", "builtin:ci", "--dry-run"}, &stdout, &stderr)
	if code != ExitOK {
		t.Fatalf("expected exit code %d, got %d. stderr: %s", ExitOK, code, stderr.String())
	}
	out := stdout.String()
	if !strings.Contains(out, "doctor") {
		t.Fatalf("expected doctor step, got: %s", out)
	}
	if !strings.Contains(out, "examples_verify") {
		t.Fatalf("expected examples_verify step, got: %s", out)
	}
	if !strings.Contains(out, "verify_ci_standard") {
		t.Fatalf("expected verify_ci_standard step, got: %s", out)
	}
}

func TestRunDryRunJSON(t *testing.T) {
	var stdout bytes.Buffer
	var stderr bytes.Buffer
	code := Run([]string{"run", "builtin:ci", "--dry-run", "--json"}, &stdout, &stderr)
	if code != ExitOK {
		t.Fatalf("expected exit code %d, got %d. stderr: %s", ExitOK, code, stderr.String())
	}
	var result RunResult
	if err := json.Unmarshal(stdout.Bytes(), &result); err != nil {
		t.Fatalf("expected valid JSON: %v\noutput: %s", err, stdout.String())
	}
	if result.Recipe != "builtin:ci" {
		t.Fatalf("expected recipe builtin:ci, got: %s", result.Recipe)
	}
	if !result.OK {
		t.Fatalf("expected ok=true, got: %+v", result)
	}
	if len(result.Steps) == 0 {
		t.Fatalf("expected steps, got: %+v", result)
	}
}

func TestRunUnknownRecipe(t *testing.T) {
	var stdout bytes.Buffer
	var stderr bytes.Buffer
	code := Run([]string{"run", "unknown"}, &stdout, &stderr)
	if code != ExitUsage {
		t.Fatalf("expected exit code %d, got %d. stderr: %s", ExitUsage, code, stderr.String())
	}
	if !strings.Contains(stderr.String(), "unknown recipe") {
		t.Fatalf("expected unknown recipe error, got: %q", stderr.String())
	}
}

func TestRunHelp(t *testing.T) {
	var stdout bytes.Buffer
	var stderr bytes.Buffer
	code := Run([]string{"run", "--help"}, &stdout, &stderr)
	if code != ExitUsage {
		t.Fatalf("expected exit code %d, got %d", ExitUsage, code)
	}
	if !strings.Contains(stderr.String(), "usage:") {
		t.Fatalf("expected usage message, got: %q", stderr.String())
	}
}

func TestRunInHelpListing(t *testing.T) {
	var stdout bytes.Buffer
	var stderr bytes.Buffer
	code := Run([]string{"--help"}, &stdout, &stderr)
	if code != ExitOK {
		t.Fatalf("expected exit code %d, got %d", ExitOK, code)
	}
	if !strings.Contains(stdout.String(), "run") {
		t.Fatalf("expected help to include run, got: %s", stdout.String())
	}
}

func TestRunMaturityBeta(t *testing.T) {
	var stdout bytes.Buffer
	var stderr bytes.Buffer
	code := Run([]string{"--help-maturity"}, &stdout, &stderr)
	if code != ExitOK {
		t.Fatalf("expected exit code %d, got %d", ExitOK, code)
	}
	out := stdout.String()
	if !strings.Contains(out, "run") {
		t.Fatalf("expected --help-maturity to include run, got: %s", out)
	}
	betaIdx := strings.Index(out, "beta:")
	expIdx := strings.Index(out, "experimental:")
	// Look for the line that starts with "  run" inside the beta block
	runIdx := strings.Index(out, "\n  run")
	if betaIdx == -1 || expIdx == -1 || runIdx == -1 {
		t.Fatalf("missing expected sections")
	}
	if runIdx < betaIdx || runIdx > expIdx {
		t.Fatalf("expected run to appear under beta section")
	}
}
